engine/src/main/java/org/hibernate/validator/internal/engine/valueextraction/ValueExtractorResolver.java

```diff
@@ -0,0 +1,331 @@
+/*
+ * Hibernate Validator, declare and validate application constraints
+ *
+ * License: Apache License, Version 2.0
+ * See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+package org.hibernate.validator.internal.engine.valueextraction;
+
+import java.lang.invoke.MethodHandles;
+import java.lang.reflect.Type;
+import java.lang.reflect.TypeVariable;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+import javax.validation.valueextraction.ValueExtractor;
+
+import org.hibernate.validator.internal.util.CollectionHelper;
+import org.hibernate.validator.internal.util.TypeHelper;
+import org.hibernate.validator.internal.util.TypeVariableBindings;
+import org.hibernate.validator.internal.util.TypeVariables;
+import org.hibernate.validator.internal.util.logging.Log;
+import org.hibernate.validator.internal.util.logging.LoggerFactory;
+
+/**
+ * Contains resolving algorithms for {@link ValueExtractor}s, and caches for these
+ * extractors based on container types.
+ *
+ * @author Gunnar Morling
+ * @author Guillaume Smet
+ * @author Marko Bekhta
+ */
+class ValueExtractorResolver {
+
+	private static final Log LOG = LoggerFactory.make( MethodHandles.lookup() );
+
+	private final ConcurrentHashMap<ValueExtractorCacheKey, Set<ValueExtractorDescriptor>> possibleValueExtractorsByRuntimeTypeAndTypeParameter = new ConcurrentHashMap<>();
+	private final ConcurrentHashMap<Class<?>, Set<ValueExtractorDescriptor>> possibleValueExtractorsByRuntimeType = new ConcurrentHashMap<>();
+	private final Set<Class<?>> nonContainerTypes = new HashSet<>();
+
+	private final List<ValueExtractorDescriptor> valueExtractors;
+
+	ValueExtractorResolver(ValueExtractorManager valueExtractorManager) {
+		this.valueExtractors = valueExtractorManager.getValueExtractors().values().stream().collect(
```

Comment:
I would say let's pass the list here rather than having a dependency back to `ValueExtractorManager`.